core/packages/Condensation.cc

```diff
@@ -13,94 +13,123 @@ Condensation::Node &Condensation::pushNode(ImportType type) {
     return node;
 }
 
-const Condensation::Traversal Condensation::computeTraversal(const core::GlobalState &gs) const {
-    Traversal result;
+namespace {
 
-    // The nodes we're currently exploring, and will explore on the next iteration.
-    vector<uint32_t> frontier;
-    vector<uint32_t> next;
+struct TraversalBuilder {
+    absl::Span<const Condensation::Node> nodes;
+    vector<uint32_t> remainingImports;
+    vector<vector<uint32_t>> backEdges;
 
-    vector<int> remainingImports(this->nodes_.size(), 0);
-    vector<vector<uint32_t>> backEdges(this->nodes_.size());
+    vector<uint32_t> sccLengths;
+    vector<uint32_t> stratumLengths;
 
-    // Seed the needed imports for the traversal from the roots.
-    for (auto &node : this->nodes_) {
-        remainingImports[node.id] = node.imports.size();
+    Condensation::Traversal result;
 
-        if (node.imports.empty()) {
-            frontier.emplace_back(node.id);
-        }
+    TraversalBuilder(const core::GlobalState &gs, absl::Span<const Condensation::Node> nodes)
+        : nodes{nodes}, remainingImports(this->nodes.size(), 0), backEdges(this->nodes.size()) {
+        result.packages.reserve(gs.packageDB().packages().size());
+    }
+
+    // Return the set of all root packages, for the first stratum in the traversal.
+    vector<uint32_t> roots() {
+        vector<uint32_t> frontier;
+
+        // Seed the needed imports for the traversal from the roots.
+        for (auto &node : this->nodes) {
+            this->remainingImports[node.id] = node.imports.size();
```

Comment:
unrelated: one thing that came up while I was doing the work to shrink packages: `sccID_` being an `int` with `-1` as a special value was kind of scary, because I wasn't sure which ints were actually SCC IDs. Maybe we want a newtype wrapper for SCC IDs?

Reply:
That's not a bad idea, it would be nice to not accidentally confuse the id with another int value.